runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -6738,8 +6751,16 @@ static void genHeapAlloc(TR::Node *node, TR::Instruction *&iCursor, TR_OpaqueCla
          iCursor = generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, zeroReg, 0, iCursor);
          if (sizeInReg)
             {
-            iCursor = generateTrg1ImmInstruction(cg, TR::InstOpCode::lis, node, sizeReg, (allocSize >> 16) + ((allocSize & (1 << 15)) ? 1 : 0), iCursor);
-            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, sizeReg, sizeReg, allocSize & 0x0000FFFF, iCursor);
+            if (0x00008000 == HI_VALUE(allocSize))
+               {
+               iCursor = generateTrg1ImmInstruction(cg, TR::InstOpCode::lis, node, sizeReg, HI_VALUE(allocSize), iCursor);
+               iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, sizeReg, sizeReg, 0, 0xFFFFFFFF, iCursor);
```

Comment:
Could you please clarify why this instruction is needed? I thought just lis would be enough for any HI_VALUE.

Reply:
I made a mistake there. Those two lines are just wrong. They should be a `lis` instruction followed by an `addis` instruction with the intent of putting `0x000000008000000` into a register.

Reply:
I think we could've just called lis with 0xFFFF8000. This is just to avoid the ASSERT but the register will end up with the same bit pattern.

Reply:
No, we can't do that. As I pointed out previously in https://github.com/eclipse/omr/pull/5213#discussion_r428859832, the assert would not be a false positive in these instances. When using the full 64-bit result (as is done here when running on 64-bit systems), using `-0x8000` instead of `0x8000` will not end up with the correct bit pattern in the register and will result in incorrect behaviour.

Reply:
Right! Forgot that lis is the same as addis with 0.